modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java

```diff
@@ -3424,17 +3417,38 @@ else if (X.hasCause(e, SocketTimeoutException.class))
                     errs.addSuppressed(new IgniteCheckedException("Failed to connect to address " +
                         "[addr=" + addr + ", err=" + e.getMessage() + ']', e));
 
-                    // Reconnect for the second time, if connection is not established.
-                    if (!failureDetThrReached && connectAttempts < 5 &&
-                        (X.hasCause(e, ConnectException.class, HandshakeException.class, SocketTimeoutException.class))) {
-                        U.sleep(200);
+                    if (X.hasCause(e, ConnectException.class, HandshakeException.class, SocketTimeoutException.class) &&
+                        !failureDetThrReached && attempt < reconCnt) {
 
-                        connectAttempts++;
+                        // Wait a bit before next retry to mitigate short network or node problems.
+                        long delay = failureDetectionTimeoutEnabled() ?
+                            timeoutHelper.remainingTime(U.currentTimeMillis()) / (reconCnt - attempt) :
+                            connTimeout0 - (U.currentTimeMillis() - start);
 
-                        continue;
-                    }
+                        // Reconnect again if connection was not established within current timeout chunk.
+                        attempt++;
 
-                    break;
+                        if (delay > 0 && delay < 200) delay = 200;
+
+                        while (delay > 0) {
```

Comment:
Now, thread can spent more than 200ms here. During that time node will reject incoming connection.